feat: nydus support encrypted images

taoohong opened the pull request:
Extend native nydus v6 to support handling encrypted containers images:
* An encrypted nydus image is composed of encrypted bootstrap and chunk-level encrypted data blobs. The bootstrap is encrypted by the Ocicrypt and the data blobs are encrypted by aes-128-xts with randomly generated key and iv at chunk-level.
* The chunk encryption key and iv are stored in the blob info reusing some items of the structure to save reserved space.
* For every data blob, all the chunk data, conpression context. table and compression context table header are encrypted.
* Encrypted chunk data will be decrypted first, and then be decompressed while bing fetched by the storage backend.

Related [issue](https://github.com/dragonflyoss/image-service/issues/1334)

jiangliu commented on utils/src/crypt.rs:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2022 Alibaba Cloud. All rights reserved.
+// Copyright (C) 2023 Alibaba Cloud. All rights reserved.
```
Should it be 2022-2023?

taoohong replied:
done

jiangliu commented on builder/src/compact.rs:
```diff
@@ -552,6 +552,9 @@ impl BlobCompactor {
                         build_ctx.blob_features,
                         build_ctx.compressor,
                         build_ctx.digester,
+                        build_ctx.cipher,
```
Seems it helps to define a new helper `BlobContext::new_with_builid_ctx()`?

jiangliu commented on builder/src/core/blob.rs:
```diff
@@ -194,8 +195,22 @@ impl Blob {
         }
+
+        let encrypted_ci_data = if encrypt {
+            if let Some(cipher_ctx) = &blob_ctx.cipher_ctx {
```
Seems this could be abstracted as a helper.

jiangliu commented on builder/src/core/blob.rs:
```diff
@@ -212,18 +227,32 @@ impl Blob {
             }
         }
-        blob_ctx.write_data(blob_writer, header.as_bytes())?;
+        blob_ctx.write_data(blob_writer, &encrypted_header)?;
```
Could you please help to check whether the `to_vec()` at line 251 is necessary or not?

taoohong replied:
It's necessary, when compressed is false, the compressed data equals &blob_ctx.blob_meta_info.as_byte_slice().

taoohong commented:
/retest

taoohong commented:
/retest

bergwolf commented:
> The bootstrap is encrypted by the Ocicrypt and the data blobs are encrypted by aes-128-xts with randomly generated key and iv at chunk-level.

What is `iv`? And I thought we agreed on using convergent encryption for shared chunk data encryption. Any reason why it got dropped?

taoohong commented:
> > The bootstrap is encrypted by the Ocicrypt and the data blobs are encrypted by aes-128-xts with randomly generated key and iv at chunk-level.
> 
> The IV is used to ensure that the same plaintext block encrypted with the same data encryption key produces a different ciphertext block each time. 



> > The bootstrap is encrypted by the Ocicrypt and the data blobs are encrypted by aes-128-xts with randomly generated key and iv at chunk-level.
> 
> What is `iv`? And I thought we agreed on using convergent encryption for shared chunk data encryption. Any reason why it got dropped?

The IV is used to ensure that the same plaintext block encrypted with the same data encryption key produces a different ciphertext block each time.  It's usually suggested to be provided when using AES in xts mode.
About `Convergent encryption` you mean CipherContext with convergent_encryption set true? It's not really dropped, the related codes are moved to `utils/crypt.rs` to make them capable of being more commonly and easily used by storage or rafs subsystems.

bergwolf commented on rafs/src/metadata/layout/v6.rs:
```diff
@@ -1413,11 +1423,23 @@ struct RafsV6Blob {
+    // |     8bytes       |      8bytes      |
+    // +------------------+------------------+
+    //  \_         cipher_iv[0..16]        _/
+    cipher_iv: [u8; 8],
```
ok, google tells that it is an cipher's `initialization vector`. But is it needed for decryption at all? If not, we should not save it in v6 blob.

bergwolf replied:
ok, after looking at the rest of the PR, I see that IV is needed for decryption as well.

bergwolf commented on utils/src/crypt.rs:
```diff
@@ -320,6 +337,74 @@ impl Cipher {
+        let mut buf = vec![0u8; size];
+        if let Err(e) = rand::rand_bytes(&mut buf) {
+            Err(eother!(format!(
+                "failed to generate key for Aes256Xts, {}",
```
not just Aes256Xts right ?

taoohong replied:
Yeah, I will fix it. : )

bergwolf commented on utils/src/crypt.rs:
```diff
@@ -320,6 +337,74 @@ impl Cipher {
+impl CipherContext {
+    /// Create a new instance of [CipherContext].
+    pub fn new(key: Vec<u8>, iv: Vec<u8>, convergent_encryption: bool) -> Result<Self, Error> {
+        if key.len() != 32 {
```
The check and the one below, are only for Aes128Xts? If so, please rename this as `new_aes128xts`. `Cipher` is implemented to work for different cipher algorithms. It's confusing `CipherContext` only works for aes128xts. If the limitation is necessary, please make sure to highlight it with proper function names and comments.

taoohong replied:
I'll fix it. THX!

bergwolf reviewed (dismissed):
Thanks!

bergwolf commented:
@taoohong The implementation looks good. Could you add some integration tests (in the `smoke/tests/` directory) so that it is actually tested?

taoohong commented:
> Thanks!



> @taoohong The implementation looks good. Could you add some integration tests (in the `smoke/tests/` directory) so that it is actually tested?

I'd like to. THX for your review 👍

taoohong committed:
feat: nydus support encrypted images

Extend native nydus v6 to support handling encrypted
containers images:
* An encrypted nydus image is composed of encrypted
bootstrap and chunk-level encrypted data blobs. The
bootstrap is encrypted by the Ocicrypt and the data
blobs are encrypted by aes-128-xts with randomly
generated key and iv at chunk-level.
* For every data blob, all the chunk data, conpression
context. table and compression context table header
are encrypted.
* The chunk encryption key and iv are stored in the blob
info reusing some items of the structure to save reserved
space.
* Encrypted chunk data will be decrypted and then be
decompressed while be fetched by the storage backend.
* Encrypted or unencrypted blobs can be merged together.

Signed-off-by: taohong <taoohong@linux.alibaba.com>

taoohong committed:
tests: add encrypt integration test

Add image encryption test integration case to Smoke test.

Signed-off-by: taohong <taoohong@linux.alibaba.com>
